role arg spec - allow file extensions in `tasks_from`

samdoran opened the pull request:
##### SUMMARY
Lookup the spec using the basename of the last part of the path or file specified in `tasks_from`. Since `tasks_form` accepts a file or path both with and without an extension, the argument spec lookup should as well.

Fixes #75456

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/playbook/role/__init__.py`


To do:
- [x] Add tests

samdoran committed:
role arg spec - allow file extensions in task_from

Lookup the spec using the basename of the file specified in tasks_from. Since
tasks_form accepts both with and without extensions, so should the argument
spec lookup.

bcoca commented on lib/ansible/playbook/role/__init__.py:
```diff
@@ -317,6 +317,12 @@ def _prepend_validation_task(self, task_data, argspecs):
             entrypoint = self._from_files.get('tasks', 'main')
+
+            # Remove the file extension
+            i = entrypoint.rfind('.')
```
or just use splitext and compare the first item ?

sivel replied:
The other concern here is pathing.  In the original report, `tasks_from` ended up as an absolute path.

bcoca replied:
the path should always be rolepath/tasks/ (we dont allow subdirs in _from), so can use basename or just capture the value before it gets transformed

bcoca replied:
he, just tested and putting the abs path in `tasks_from ` does work, so you really cannot rely on it not being a path at one point

bcoca replied:
`tasks_from: tasks/stuff.yml` .. also works

samdoran replied:
> The other concern here is pathing. In the original report, tasks_from ended up as an absolute path.

So we essentially need the [stem](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.stem). I'll work on that, and add some tests for these cases.

bcoca replied:
cannot use pathlib until we split controller tests ..

samdoran replied:
Yup, I just really like a lot of the things it provides and it's good for inspiration.

samdoran replied:
I think I have a better solution now.. Thanks for the feedback.

samdoran committed:
Extract the stem from and use that to lookup the arg spec

This allows for relative paths, absolute paths, and files with or withou
an extension.

samdoran committed:
Add stem() function

samdoran committed:
Use stem() function for getting correct entrypoint name

Change variable name to better reflect its contents.

samdoran committed:
Add integration tests

samdoran committed:
Fix unit test for older Python versions

briantist commented on test/integration/targets/roles_arg_spec/test_path_extension.yml:
```diff
@@ -0,0 +1,32 @@
+    - name: Tasks from tasks/alternate
+      import_role:
+          name: a
+          tasks_from: taks/alternate
```
```suggestion
          tasks_from: tasks/alternate
```
typo (wondering why this didn't fail the test run though 🤔)

samdoran replied:
That is because we take the `basename()` of the value in `tasks_from`.

https://github.com/ansible/ansible/blob/74f53676739600d7f8a18087aad1b1203b5ed7ac/lib/ansible/playbook/role_include.py#L141-L146

bcoca committed:
Update test/integration/targets/roles_arg_spec/test_path_extension.yml

Co-authored-by: Brian Scholer <1260690+briantist@users.noreply.github.com>

sivel commented:
I do question whether we should have a list of candidates to try.  We might just say this isn't supported, but here it is.

Consider the following tree:

```
.
└── roles
    └── foo
        └── tasks
            ├── install.yml
            └── ubuntu
                └── install.yml
```

What about calculating the relative path based on `role_path`?  And looping from most specific to least specific?

That way if someone provided `tasks_from: ubuntu/install.yml` it could have a different entry point than `tasks_from: install.yml`?

Just a thought, based on the current solution being naive about paths.  Although this PR is an improvement from one view, it actually takes away the potential features.

Again, maybe this is just something we document as not being possible.

samdoran commented:
> That way if someone provided tasks_from: ubuntu/install.yml it could have a different entry point than tasks_from: install.yml?

In general, I do like this idea. The current behavior can be a bit surprising since it's pretty flexible/naive with its path searching.

One concern I have is how we would then match the role arg spec to the correct entry point if the stem is the same but the relative path is different. It seems like that would complicate the arg spec names and matching quite a bit.

Naming the spec `install` becomes ambiguous (though it already is). Would it need to be named based on the relative path inside `tasks/`? Something like `install` and `ubunutu_install`?

sivel commented:
> Naming the spec `install` becomes ambiguous (though it already is). Would it need to be named based on the relative path inside `tasks/`? Something like `install` and `ubunutu_install`?

I was thinking literally `install` and `ubuntu/install` to basically align with the functionality already achievable prior to this PR.

bcoca commented on lib/ansible/module_utils/splitter.py:
```diff
@@ -217,3 +219,13 @@ def unquote(data):
     return data
+
+
+def stem(path):
```
per discussion, you might want to look at is_subpath (ansible.utils.path) and also put this in common/file.py or create a common/path.py?

Veratil commented:
> I have been looking into this more and found that subdirs for `tasks_from` work with `include_role` but not `import_role`. I need to track down and reconcile that behavior.
> 
> ```
>  0.1s _load_role_data: self._from_files.get('tasks')='ubuntu/install'  # include_tasks
> 
>  0.1s _load_role_data: self._from_files.get('tasks')='install'         # import_tasks
> ```
> 
> 
> The more I think about it, I do like the flexibility of having specs match the `tasks_from` hierarchy. I'm a bit leery of how tricky it will be to implement that since the task loading/finding code is separate from the spec finding code/logic.

https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/role_include.py#L154

Remove `basename` here. I wrote some tests and tested on devel and your branch, all passed. Validation is still run on the path and will fail when it's out of the role folder.
